### ISSUE TITLE
Add direct tests for libc FS functions like "ftruncate"

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/miri#3179]
"ftruncate" (and really, most of the file system APIs) is currently only tested via the libstd functions, but not via direct libc calls. With PRs like https://github.com/rust-lang/miri/pull/3177 that add support for ftruncate on targets where libstd uses other functions (like ftruncate64), it'd be good to have direct tests as well.

**Comment by RalfJung** [comment]
Right, basically these tests should be fairly similar to [this one](https://github.com/rust-lang/miri/blob/master/tests/pass/shims/fs.rs) but directly using the libc APIs rather than calling Rust's `std::fs` functions.

You can use the `CString` type to generate C-compatible strings, no need to manually push `\0`.

**Comment by RalfJung** [comment]
Support for extra modes in `open` (as implemented in https://github.com/rust-lang/miri/pull/3240) should also be covered by such a test.

**Comment by eduardorittner** [comment]
I looked over all `foreign_item.rs`s and for each function there looked for corresponding direct libc tests, these are the functions which do not have any direct tests associated, categorized by where they are located in the source code. If I found a single instance of a direct function call in tests, I immediately discarded it, so it is possible for functions not listed here to have been called only once in a test and not have sufficient test coverage. I also did not include any functions which check for `this.frame_in_std()`.

# Shared (`foreign_items.rs`)
- [ ] `exit`
- [ ] `abort`
- [x] `memcmp`

# Unix (`unix/foreign_items.rs`)

- [x] `getenv`
- [x] `setenv`
- [x] `unsetenv`
- [ ] `getcwd`
- [ ] `chdir`
- [ ] `realpath`
- [x] `pread`
- [x] `pwrite`
- [ ] `lseek`
- [x] `fcntl`
- [x] `fstat`
- [x] `stat`
- [x] `lstat`
- [ ] `unlink`
- [ ] `symlink`
- [ ] `mkdir`
- [ ] `rmdir`
- [ ] `opendir`
- [ ] `readdir`
- [ ] `closedir`
- [ ] `fsync`
- [ ] `fdatasync`
- [ ] `aligned_alloc`
- [ ] `pthread_key_delete`
- [ ] `pthread_getspecific`
- [ ] `pthread_setspecific`
- [ ] `sched_getaffinity`
- [ ] `pthread_atfork`
- [x] `arc4random_buf`
- [x] `_Unwind_RaiseException`

## Android (`unix/android/foreign_items.rs`)

All implemented

## FreeBSD (`unix/freebds/foreign_items.rs`)

All implemented

## Linux (`unix/linux/foreign_items.rs`)

- [ ] `open64`
- [ ] `pread64`
- [ ] `pwrite64`
- [ ] `lseek64`
- [ ] `readdir64`
- [ ] `statx`
- [ ] `__xpg_strerror_r`
- [ ] `__libc_current_sigrtmin`
- [ ] `__libc_current_sigrtmax`

## MacOS (`unix/macos/foreign_items.rs`)

- [ ] `__error`
- [ ] `readdir_r`
- [ ] `_NSGetEnviron`
- [ ] `_NSGetArgc`
- [ ] `_NSGetArgv`
- [ ] `_NSGetExecutablePath`
- [ ] `mach_timebase_info`
- [ ] `_tlv_atexit`
- [ ] `pthread_get_stackaddr_np`
- [ ] `pthread_get_stacksize_np`

## Solarish (`unix/solarish/foreign_items.rs`)

- [ ] `__xnet_socketpair`
- [ ] `stack_getbounds`
- [ ] `pset_info`
- [ ] `__sysconf_xpg7`

# Windows (`windows/foreign_items.rs`)

- [ ] `GetEnvironmentVariableW`
- [ ] `SetEnvironmentVariableW`
- [ ] `GetEnvironmentStringsW`
- [ ] `FreeEnvironmentStringsW`
- [ ] `GetCurrentDirectoryW`
- [ ] `SetCurrentDirectoryW`
- [ ] `GetUserProfileDirectoryW`
- [ ] `GetCurrentProcessId`
- [ ] `GetFullPathNameW`
- [ ] `HeapAlloc`
- [ ] `HeapFree`
- [ ] `HeapFree`
- [ ] `HeapReAlloc`
- [ ] `GetSystemInfo`
- [ ] `GetCommandLineW`
- [ ] `GetSystemTimeAsFileTime`
- [ ] `QueryPerformanceCounter`
- [ ] `QueryPerformanceFrequency`
- [ ] `CreateWaitableTimerExW`
- [ ] `WaitOnAddress`
- [ ] `WaitByAddressSingle`
- [ ] `WaitByAddressAll`
- [ ] `SetThreadDescription`
- [ ] `GetThreadDescription`
- [ ] `ExitProcess`
- [ ] `ProcessPrng`
- [ ] `GetConsoleScreenBufferInfo`
- [ ] `GetStdHandle`
- [ ] `GetModuleFileNameW`
- [ ] `FormatMessageW`

If I missed any (or incorrectly counted a function as not being tested when it was), feel free to tell me

**Comment by RalfJung** [comment]
Thanks for making a list!
Your categorization is a bit off, `lseek*` and `f*sync` are not "directory operations".

And there are definitely more, e.g. some of the ones in `src/shims/unix/linux/foreign_items.rs` (such as `readdir64` and `statx`) or in `src/shims/unix/macos/foreign_items.rs` (such as `mach_timebase_info`, `_NSGetExecutablePath`).

**Comment by eduardorittner** [comment]
Thanks for the heads up, I'll try to look a bit more closely to add the ones I've missed to the list

**Comment by eduardorittner** [comment]
I took a look at `pthread_getattr_np`, there are not `libc::pthread_getattr_np` in any of the tests, but due to the comment this looks like a special case, should we still add direct libc tests for this?

```
// Incomplete shims that we "stub out" just to get pre-main initialization code to work.
// These shims are enabled only when the caller is in the standard library.
```

**Comment by RalfJung** [comment]
Yeah, the shims with this extra `this.frame_in_std()` check do not need dedicates tests.

**Comment by eduardorittner** [comment]
Another question, there are some functions (specially in linux) which start with a `__`, like `__errno`, `__errno_location` and so on. Do they need dedicated tests as well?

**Comment by bjorn3** [comment]
Those are still part of the libc ABI and actually used by user crates, so yes.
